Viewport Transitions : add pure-js TransitionManager

1chandu opened the pull request:
- Extract all transition logic into re-usable pure Javascript class 'TransitionManager'
- Keep only react specific code in `ViewportController`.
- Define clear API between `TransitionManager` and `ViewportController` for clarity and scalability.

Verified with all 3 viewport transition examples.

1chandu commented:
It looks like I also need to add support for 'viewport descriptors', I will add those changes to this PR.

ibgreen commented on examples/wip/viewport-animation-3d-heatmap/app.js:
```diff
@@ -27,9 +25,9 @@ class Root extends Component {
+      transitionDuration: 0,
       viewportToggled: false,
-      onAnimationStop: this.rotateCameara.bind(this)
+      onTransitionEnd: this.rotateCameara.bind(this)
```
Nit: typo `Camera`

ibgreen commented on src/react/experimental/transition-manager.js:
```diff
@@ -0,0 +1,209 @@
+
+/* eslint-disable max-depth */
```
Is this actually needed? Could it be used to wrap only the part that is needed?

1chandu replied:
Moved to code block where it is needed.

ibgreen commented on src/react/experimental/transition-manager.js:
```diff
@@ -0,0 +1,209 @@
+  VIEWPORT_PROPS
+} from './viewport-transition-utils';
+
+// TODO: make the frequency customizalbe , adapt to frame rate
```
Nit type customizable

ibgreen commented on src/react/experimental/transition-manager.js:
```diff
@@ -0,0 +1,209 @@
+      endViewport: null
+    };
+
+    this._updateViewport = this._updateViewport.bind(this);
```
Normally only callbacks need to be bound. Maybe review these and see if can omit binding some?

1chandu replied:
Yep, I am not passing them around, `bind` not needed, removed.

ibgreen commented on src/react/experimental/transition-manager.js:
```diff
@@ -0,0 +1,209 @@
+  }
+
+  processViewportChange(props, nextProps) {
+    // NOTE: Be cautious re-ordering statements in this function.
```
Maybe add a few guiding comments inside the function?

1chandu replied:
Added.

ibgreen commented on src/react/experimental/transition-manager.js:
```diff
@@ -0,0 +1,209 @@
+    return this.transitionContext.viewport;
+  }
+
+  processViewportChange(props, nextProps) {
```
Add a one line header over each function explaining its function

ibgreen commented on src/react/experimental/transition-manager.js:
```diff
@@ -0,0 +1,209 @@
+
+  // Helper methods
+
+  _areViewportsEqual(startViewport, endViewport) {
```
oldViewports, newViewports?

1chandu replied:
Each argument only contains all params of one viewport.

ibgreen commented on src/react/experimental/transition-manager.js:
```diff
@@ -0,0 +1,209 @@
+    return props.transitionDuration !== 0;
+  }
+
+  _isTheUpdateDueToCurrentTransition(nextProps) {
```
Nit: Drop `The`: `_isUpdateDue...`

ibgreen commented on src/react/experimental/viewport-transition-utils.js:
```diff
@@ -4,30 +4,50 @@ import {projectFlat, unprojectFlat} from 'viewport-mercator-project';
+export const VIEWPORT_PROPS = ['longitude', 'latitude', 'zoom', 'bearing', 'pitch',
+  'position', 'width', 'height', 'minZoom', 'maxZoom'];
+
+export function extractViewportFrom(params) {
```
`extractViewportFrom(props)`? Instead of `params`?

ibgreen replied:
Handle `viewports` prop?

ibgreen commented on src/react/viewport-controller.js:
```diff
@@ -87,24 +129,69 @@ export default class ViewportController extends PureComponent {
     if (this._controls) {
       this._controls.setOptions(nextProps);
     }
+    if (this._transitionManger) {
```
Shouldn't it read all props, not just viewports?

1chandu replied:
`processViewportChange` takes all props, extracts viewport props from it and processes only a viewport change. In this PR I am not handling `viewports` (multi-viewport prop), that will be in next PR.

1chandu commented:
@ibgreen @Pessimistress @jianhuang01 , addressed all comments. Also disabled partial multi-viewport transition in `ViewportController`,  transition is only supported when viewport is specified in base level props (not when using `viewport` or `viewports` prop), I can do multi-viewport transition in next PR, which requires supporting `viewport descriptors` and more careful testing of  transitioned viewport to various `ViewState` conversions.

Pessimistress commented on src/react/experimental/transition-manager.js:
```diff
@@ -0,0 +1,205 @@
+      let newTime = time + delta;
+      // Make sure interplation step is always ends with time = 1.0
+      if (newTime > 1.0) {
+        newTime = 1.0 - delta;
```
`newTime = 1.0` ?

Pessimistress commented on src/react/experimental/transition-manager.js:
```diff
@@ -0,0 +1,205 @@
+      this.transitionContext.endViewport,
+      easing
+    );
+    assert(time <= 1.0);
```
This may fail if `delta > 1`. At least provide an error message.

Pessimistress commented on src/react/experimental/viewport-transition-utils.js:
```diff
@@ -4,30 +4,46 @@ import {projectFlat, unprojectFlat} from 'viewport-mercator-project';
 import {Vector2} from 'math.gl';
 
 const EPSILON = 0.01;
+export const VIEWPORT_PROPS = ['longitude', 'latitude', 'zoom', 'bearing', 'pitch',
```
Why interpolate `minZoom` and `maxZoom`? It may cause weird zoom output during interpolation.
If it's deemed necessary, should we add `minPitch` and `maxPitch` too?

Pessimistress commented on src/react/experimental/viewport-transition-utils.js:
```diff
@@ -4,30 +4,46 @@ import {projectFlat, unprojectFlat} from 'viewport-mercator-project';
   }
 
-  for (const p of ['longitude', 'latitude', 'zoom', 'bearing', 'pitch']) {
+  for (const p of ['longitude', 'latitude', 'zoom', 'bearing', 'pitch', 'position']) {
```
Define constant
